rp2040-hal/src/adc.rs

```diff
@@ -153,3 +154,210 @@ where
+        Fifo { adc: self.adc }
+    }
+
+    /// Enable ADC fifo with DMA and start free-running conversion
+    ///
+    /// In addition to being usable the same way as a `Fifo` returned from `start_fifo`,
+    /// the Fifo returned by this function can also be used as a source for DMA transfers.
+    pub fn start_fifo_with_dma(self, thresh: u8) -> Fifo<'a, true> {
+        self.adc.device.fcs.modify(|_, w| unsafe {
+            w.en().set_bit()
+                .dreq_en().set_bit()
+                .thresh().bits(thresh)
+        });
+        self.adc.device.cs.modify(|_, w| w.start_many().set_bit());
+        Fifo { adc: self.adc }
+    }
+}
+
+/// Represents the ADC fifo, when used in free running mode
+///
+/// The fifo can be used in one of two ways:
+/// 1. Directly, by polling `len()` and calling `read()`:
+/// ```
+/// let fifo = adc.free_running().sample_rate(1).initial_input(&mut adc_pin).start_fifo();
+/// loop {
+///   // wait for fifo to fill the first value:
+///   if fifo.len() > 0 {
+///     // read it & print it!
+///     println!("Sample: {}", fifo.read());
+///   }
+/// }
+/// ```
+///
+pub struct Fifo<'a, const DmaEnabled: bool> {
+    adc: &'a mut Adc,
+}
+
+impl<'a, const DmaEnabled: bool> Fifo<'a, DmaEnabled> {
+    pub fn clock_divider(&mut self) -> (u16, u8) {
+        let r = self.adc.device.div.read();
+        (r.int().bits(), r.frac().bits())
+    }
+
+    /// Returns the number of elements currently in the fifo
+    pub fn len(&mut self) -> u8 {
+        self.adc.device.fcs.read().level().bits()
+    }
+
+    /// Check if there was a fifo overrun
+    ///
+    /// An overrun happens when the fifo is filled up faster than `read` is called to consume it.
+    ///
+    /// This function also clears the `over` bit if it was set.
+    pub fn is_over(&mut self) -> bool {
+        let over = self.adc.device.fcs.read().over().bit();
+        if over {
+            self.adc.device.fcs.modify(|_, w| w.over().set_bit());
+        }
+        over
+    }
+
+    /// Check if there was a fifo underrun
+    ///
+    /// An underrun happens when `read` is called on an empty fifo (`len() == 0`).
+    ///
+    /// This function also clears the `under` bit if it was set.
+    pub fn is_under(&mut self) -> bool {
+        let under = self.adc.device.fcs.read().under().bit();
+        if under {
+            self.adc.device.fcs.modify(|_, w| w.under().set_bit());
+        }
+        under
+    }
+
+    /// Read a single value from the fifo
+    pub fn read(&mut self) -> u16 {
+        self.adc.device.fifo.read().val().bits()
+    }
+
+    /// Stop capturing in free running mode.
+    ///
+    /// Disables the ADC fifo and stops ADC capture.
+    /// Returns the underlying `Adc` instance.
+    pub fn stop(self) -> &'a mut Adc {
+        self.adc.device.cs.modify(|_, w| w.start_many().clear_bit());
+        self.adc.device.fcs.modify(|_, w| w.en().clear_bit());
+        self.adc
+    }
+}
```

Comment:
Just an observation, not sure if it's good or bad: This doesn't disable DREQ. I _think_ it doesn't do harm to keep it enabled, and I also _think_ it would be fine to disable it here. Anyone with deeper knowledge of the DMA engine around who could tell if one or the other would be better?

Reply:
True. I don't know what's the best practice here.

We could store the register values when the `Fifo` is constructed, and reset them to those exact values when `stop`ping.

Reply:
Update: turns out we also need to drain the FIFO here. Otherwise if there is still a pending interrupt, the ISR will still be called, even if interrupts are already disabled at that point.